File: workers/cad/flatcraft_cad/validate/perforation.py
"""Parity perforation-валідатор (дзеркало TS validators/perforation.ts).

Геометрична коректність grid отворів перфо-панелі: крок (pitch) має бути
більший за розмір отвору по кожній осі, інакше сусідні отвори торкаються/
перетинаються і зливаються у суцільний проріз (а BOM рахує їх як N окремих
отворів — невідповідність заготовки).

Правило (паритет з TS):
    pitch > hole_size   → валідно
    pitch <= hole_size  → HOLES_OVERLAP

де hole_size — сторона квадрата (perforated_panel_square) або діаметр
(perforated_panel). Остання лінія оборони у воркері — навіть якщо API-gate
обійдено, відмова ДО CAD-операції/запису у R2 з 422 RFC 9457.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

_AXIS_LABEL = {
    "pitch_x_mm": "Крок X",
    "pitch_y_mm": "Крок Y",
}


@dataclass(frozen=True)
class PerforationError:
    """Один елемент 422-detail. `code` збігається з кодом TS-валідатора."""

    code: str
    field: str
    detail: str


def _fmt(n: float) -> str:
    return f"{round(n, 3):g}"


def _overlap_msg(which: str, pitch: float, hole_size: float, shape_word: str) -> str:
    return (
        f"Збільшіть «{_AXIS_LABEL.get(which, which)}»: крок {_fmt(pitch)} мм має бути "
        f"більший за {shape_word} отвору {_fmt(hole_size)} мм — інакше сусідні отвори "
        "перетинаються."
    )


def validate_perforation(
    template_slug: str,
    parameters: dict[str, Any],
) -> list[PerforationError]:
    """Перевіряє grid перфорації. Порожній список → валідно.

    Невідомий slug або невизначений/непозитивний розмір отвору → [] (останнє
    відсіється Pydantic-діапазонами окремо).
    """
    if template_slug == "perforated_panel_square":
        hole_size = parameters.get("hole_size_mm")
        shape_word = "сторону"
    elif template_slug == "perforated_panel":
        hole_size = parameters.get("hole_diameter_mm")
        shape_word = "діаметр"
    else:
        return []

    if hole_size is None or hole_size <= 0:
        return []

    errors: list[PerforationError] = []
    for which in ("pitch_x_mm", "pitch_y_mm"):
        pitch = parameters.get(which)
        if pitch is None:
            continue
        # Валідно при pitch > hole_size. pitch <= hole_size → торкання/перетин.
        if pitch <= hole_size:
            errors.append(
                PerforationError(
                    code="HOLES_OVERLAP",
                    field=which,
                    detail=_overlap_msg(which, float(pitch), float(hole_size), shape_word),
                )
            )
    return errors


def validate_export_perforation(
    template_slug: str,
    parameters: dict[str, Any],
) -> list[dict[str, str]]:
    """JSON-готовий список dict-ів для 422-detail (паритет з validate_export)."""
    return [asdict(e) for e in validate_perforation(template_slug, parameters)]

File: packages/cad-engine/src/validators/perforation.ts
/**
 * Perforation-валідатор: геометрична коректність grid отворів перфо-панелі.
 *
 * Окремий концерн від `validateProfile` (той — про достатність плеча/полиці
 * відносно гибу, дзеркало geometry.ts проти крашу R3F). Тут — про наповненість
 * робочої площини отворами: якщо крок (pitch) не більший за розмір отвору,
 * сусідні отвори перетинаються і зливаються у суцільний проріз — заготовка
 * фізично інша за те, що рахує BOM («N окремих отворів»).
 *
 * Правило: для НЕперетину потрібен додатний місток (ligament) між отворами:
 *   pitch > hole_size   (валідно)
 *   pitch <= hole_size  (отвори торкаються/перетинаються → invalid)
 * де hole_size = `hole_size_mm` (сторона квадрата при hole_shape='square' або
 * діаметр при 'circle'). Перевірка по кожній осі окремо (pitch_x, pitch_y).
 *
 * Browser-safe: жодного `node:*`, чиста функція. Дзеркалиться у Python
 * (`workers/cad/flatcraft_cad/validate/perforation.py`) для defense-in-depth:
 * клієнт (UX-банер) + Fastify-gate + cad-worker (остання лінія) — ADR-019/026.
 */
export type PerforationIssueCode = "HOLES_OVERLAP";

export interface PerforationIssue {
  readonly code: PerforationIssueCode;
  /** Поле з проблемою: `pitch_x_mm` або `pitch_y_mm`. */
  readonly which: string;
  /** Поріг, який крок має перевищити (= розмір отвору, мм). */
  readonly min: number;
  /** Введений крок (мм). */
  readonly got: number;
  /** Дружнє україномовне повідомлення (підказка, як виправити). */
  readonly message: string;
}

export interface PerforationValidationInput {
  readonly templateSlug: "perforated_panel";
  readonly parameters: Readonly<Record<string, number | string>>;
}

const AXIS_LABEL: Record<string, string> = {
  pitch_x_mm: "Крок X",
  pitch_y_mm: "Крок Y",
};

/** Округлення до 0.001 мм без хвостових нулів — для повідомлень. */
function fmt(n: number): string {
  return String(Math.round(n * 1000) / 1000);
}

function overlapMessage(which: string, pitch: number, holeSize: number, shapeWord: string): string {
  return (
    `Збільшіть «${AXIS_LABEL[which] ?? which}»: крок ${fmt(pitch)} мм має бути більший за ` +
    `${shapeWord} отвору ${fmt(holeSize)} мм — інакше сусідні отвори перетинаються.`
  );
}

/**
 * Валідує grid перфорації. Повертає [] коли валідно (або коли розмір отвору
 * невідомий/непозитивний — це відсіється Zod-діапазонами окремо).
 */
export function validatePerforation(input: PerforationValidationInput): PerforationIssue[] {
  const p = input.parameters;
  // Уніфікований шаблон (ADR-031): hole_size = hole_size_mm; слово у повідомленні
  // («сторону»/«діаметр») — за hole_shape.
  const holeSize = typeof p.hole_size_mm === "number" ? p.hole_size_mm : undefined;
  const shapeWord = p.hole_shape === "circle" ? "діаметр" : "сторону";

  if (holeSize === undefined || holeSize <= 0) return [];

  const issues: PerforationIssue[] = [];
  const check = (which: "pitch_x_mm" | "pitch_y_mm"): void => {
    const pitch = p[which];
    if (typeof pitch !== "number") return;
    // Валідно при pitch > holeSize. pitch <= holeSize → торкання/перетин.
    if (pitch <= holeSize) {
      issues.push({
        code: "HOLES_OVERLAP",
        which,
        min: holeSize,
        got: pitch,
        message: overlapMessage(which, pitch, holeSize, shapeWord),
      });
    }
  };
  check("pitch_x_mm");
  check("pitch_y_mm");
  return issues;
}
